src/Workspaces/Core/Portable/Execution/Serialization/AssetBuilder.cs

```diff
@@ -0,0 +1,204 @@
+        }
+
+        private Task<Asset> CreateParseOptionsAsync(Project project, string kind, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            return Task.FromResult<Asset>(new LanguageSpecificAsset<ParseOptions>(project.Language, project.ParseOptions, kind, _serializer.Serialize));
+        }
+
+        private Task<Asset> CreateProjectReferenceAsync(ProjectReference reference, string kind, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            return Task.FromResult<Asset>(new Asset<ProjectReference>(reference, kind, _serializer.Serialize));
+        }
+
+        private Task<Asset> CreateMetadataReferenceAsync(MetadataReference reference, string kind, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var checksum = _serializer.HostSerializationService.CreateChecksum(reference, cancellationToken);
+            return Task.FromResult<Asset>(new MetadataReferenceAsset(_serializer, reference, checksum, kind));
+        }
+
+        private Task<Asset> CreateAnalyzerReferenceAsync(AnalyzerReference reference, string kind, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var checksum = _serializer.HostSerializationService.CreateChecksum(reference, cancellationToken);
+            return Task.FromResult<Asset>(new AnalyzerReferenceAsset(_serializer, reference, checksum, kind));
+        }
+
+        private async Task<Asset> CreateSourceTextAsync(TextDocumentState state, string kind, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var text = await state.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var checksum = new Checksum(text.GetChecksum());
+
+            return new SourceTextAsset(_serializer, state, checksum, kind);
+        }
+
+        private SolutionChecksumObjectInfo GetInfo(Solution solution)
+        {
+            return new SolutionChecksumObjectInfo(solution.Id, solution.Version, solution.FilePath);
+        }
+
+        private ProjectChecksumObjectInfo GetInfo(Project project)
+        {
+            return new ProjectChecksumObjectInfo(project.Id, project.Version, project.Name, project.AssemblyName, project.Language, project.FilePath, project.OutputFilePath);
+        }
+
+        private DocumentChecksumObjectInfo GetInfo(TextDocument document)
+        {
+            // we might just split it to TextDocument and Document.
+            return new DocumentChecksumObjectInfo(document.Id, document.Name, document.Folders, GetSourceCodeKind(document), document.FilePath, IsGenerated(document));
+        }
+
+        private bool IsGenerated(TextDocument document)
+        {
+            var source = document.State as DocumentState;
```

Comment:
Because this code shouldn't have to know about the implementation hierarchy of TextDocumentState.  Because nwo if we introduce other subclasses and whatnot, we'll have to know to go update this code.  It's a brittle coupling problem.